R/pmcmc.R

```diff
@@ -0,0 +1,428 @@
+##' log likelihood estimate.
+##' 
+##' The pMCMC sampler then proceeds as follows, for n_mcmc iterations:
+##' At each loop iteration the pMCMC sampler performs three steps:
+##'   1. Propose new candidate samples for beta and start_date based on
+##'     the current samples, using the proposal distribution 
+##'     (currently i.i.d Gaussian with user-input sd pars_sd, and reflecting boundaries)
+##'   2. Calculate the log prior of the proposed parameters, 
+##'      Use the particle filter to estimate log likelihood of the data given the proposed parameters, as described above,
+##'      as well as proposing a model state space.
+##'      Add the log prior and log likelihood estimate to estimate the log posterior of the proposed parameters given the data.
+##'   3. Metropolis-Hastings step: The joint canditate sample (consisting of the proposed parameters 
+##'      and state space) is then accepted with probability min(1, a), where the acceptance ratio is
+##'      simply the ratio of the posterior likelihood of the proposed parameters to the posterior likelihood
+##'      of the current parameters. Note that by choosing symmetric proposal distributions by including
+##'      reflecting boundaries, we avoid the the need to include the proposal likelihood in the MH ratio.
+##'      
+##'   If the proposed parameters and states are accepted then we update the current parameters and states
+##'   to match the proposal, otherwise the previous parameters/states are retained for the next iteration.
+##'   
+##' @export
+##' @import coda
+pmcmc <- function(data,
+                  n_mcmc, 
+                  model_params = NULL,
+                  pars_obs = NULL,
+                  pars_init = NULL,
+                  pars_min = NULL,
+                  pars_max = NULL,
+                  pars_sd = NULL,
+                  pars_discrete = NULL,
+                  log_likelihood = NULL,
+                  log_prior = NULL,
+                  n_particles = 1e2,
+                  steps_per_day = 4
+                  ) {
+  
+  if (is.null(model_params)) {
+    model_params <- generate_parameters(
+      transmission_model = "POLYMOD",
+      progression_groups = list(E = 2, asympt = 1, mild = 1, ILI = 1, hosp = 2, ICU = 2, rec = 2),
+      gammas = list(E = 1/2.5, asympt = 1/2.09, mild = 1/2.09, ILI = 1/4, hosp = 2/1, ICU = 2/5, rec = 2/5),
+      beta = 0.1,
+      beta_times = "2020-01-01",
+      hosp_transmission = 0,
+      ICU_transmission = 0,
+      trans_profile = 1,
+      trans_increase = 1,
+      dt = 1/steps_per_day
+    )
+  } else {
+    if (length(model_params$beta_y)) {
+      stop("Time-varying beta not yet supported by pmcmc_sampler")
+    }
+  }
+  
+  if(is.null(pars_obs)) {
+    pars_obs <- list(phi_ICU = 0.95,
+                     k_ICU = 2,
+                     phi_death = 926 / 1019,
+                     k_death = 2,
+                     exp_noise = 1e6
+    )
+  }
+  
+  if(is.null(pars_init)) {
```

Comment:
Can these defaults for the beta ranges instead be provided as default function arguments above, rather than having everything as `NULL`. I think this format makes sense for the more complex arguments/functionally generated arguments above, but for this simpler case having them above makes it a bit easier to see where they are if changed in the future. I think doing this with `pars_obs` would also make sense, as that's more consistent with other parts of the code.

If you have time, quick checks that these are in a reasonable range would be useful, but I am happy to do this if helpful

Reply:
default function arguments passed as requested and pushed
leaving the conversation unresolved for now so I remember to add reasonable range checks

Reply:
added in reasonable range checks and pushed